src/main/java/com/minecolonies/coremod/client/gui/WindowResearchTree.java

```diff
@@ -74,57 +99,94 @@ public WindowResearchTree(final String branch, final BuildingUniversity.View bui
+            {
+                // this side won't actually start research; it'll be overridden the next colony update from the server.
+                // It will, however, update for the next WindowResearchTree if the colony update is slow to come back.
+                research.startResearch(building.getColony().getResearchManager().getResearchTree());
+                Network.getNetwork().sendToServer(new TryResearchMessage(building, research.getId(), research.getBranch(), false));
+                close();
+            }
+            else if (localResearch != null)
+            {
+                if (localResearch.getState() == ResearchState.IN_PROGRESS)
+                {
+                    drawUndoProgressButton(button);
+                }
+                if (localResearch.getState() == ResearchState.FINISHED)
+                {
+                    if(hasMax && research.getDepth() == MAX_DEPTH && building.getBuildingLevel() == building.getBuildingMaxLevel() && !research.isImmutable())
+                    {
+                        drawUndoCompleteButton(button);
+                    }
+                    for(String childId : research.getChildren())
+                    {
+                        if(building.getColony().getResearchManager().getResearchTree().getResearch(branch, childId) != null
+                          && building.getColony().getResearchManager().getResearchTree().getResearch(branch, childId).getState() != ResearchState.NOT_STARTED)
+                        {
+                            return;
+                        }
+                    }
+                    String parentId = IGlobalResearchTree.getInstance().getResearch(branch, research.getId()).getParent();
+                    while(!parentId.isEmpty())
+                    {
+                        if (IGlobalResearchTree.getInstance().getResearch(branch, parentId) != null
+                              && IGlobalResearchTree.getInstance().getResearch(branch, parentId).hasOnlyChild())
+                        {
+                            drawUndoCompleteButton(button);
+                            break;
+                        }
+                        parentId = IGlobalResearchTree.getInstance().getResearch(branch, parentId).getParent();
+                    }
+                }
+            }
+        }
+
+        if (button.getID().contains("undo."))
+        {
+            final ILocalResearch cancelResearch = building.getColony().getResearchManager().getResearchTree().getResearch(branch, button.getID().split("\\.")[1]);
+            if (cancelResearch != null)
+            {
+                // Canceled research will eventually be removed from the local tree on synchronization from server,
+                // But this can be long enough to be confusing.  NOT_STARTED isn't the same thing,
+                // but will render closely enough to prevent some confusion /or/ accidentally restarting.
```

Comment:
This sounds a bit weird handling to me

Reply:
There's logic on the server side that prevents clients issuing bad requests from resulting in lost items, so I can get rid of it if we don't want munging of the localResearchTree between updates from the server.  It just seemed clearer a way to show the research as removed while waiting for the next colony.markDirty() result to come in -- I've seen ~5 seconds on a same-machine dedicated server, and I'm sure it could be longer.

NOT_STARTED without nulling it prevents people from unintentionally trying to issue a begin research command, which may be part of why it's a bit hacky.